Docs cleanups, clarifications, and fixes (#334)
akka/akka-management: docs/src/main/paradox/akka-management.md

<a id="akka-management"></a>
# Akka Management

Akka Management is the core module of the management utilities that provides a central HTTP endpoint for Akka

> management extensions.

Reviewer: I might use ```which provides``` versus that.

## Dependencies

Akka management requires Akka 2.5 or later.

The main Akka Management dependency is called `akka-management`. By itself however it does not provide any capabilities,
and you have to combine it with the management extension libraries that you want to make use of (e.g. cluster http management,

> or cluster bootstrap). This design choice was made to make users able to include only the minimal set of features that they

Reviewer: ```suggestion
or cluster bootstrap). This design choice enables users to include only the minimal set of features they
```

:   ```java
    AkkaManagement.get(system).start();
    ```

> This is in order to allow users to prepare anything else they might want to prepare or start before exposing routes for

Reviewer: ```suggestion
This allows users to prepare anything further before exposing routes for
```

@@@ note

> HTTPS is not enabled by default as additional configuration from the developer is required This module does not provide security by default.

Reviewer 2: missing '.' after 'required'

Reviewer 3: @raboof try out the new `Insert a suggestion` in GH UI. :-)

Reviewer 3: ```suggestion
HTTPS is not enabled by default as additional configuration from the developer is required. This module does not provide security by default.
```

> HTTPS is not enabled by default as additional configuration from the developer is required This module does not provide security by default.

Reviewer: ```suggestion
HTTPS is not enabled by default, as additional configuration from the developer is required. This module does not provide security by default.
```

> It's the developer's choice to add security to this API. It is generally advisable not to expose management endpoints

Reviewer: ```suggestion
It is the developer's choice to add security to this API, and when. If enabled, it is generally advisable not to expose management endpoints
```

> on the internet.

Reviewer: Sounds very arpanet, might we say something indicating publicly or not on a private network vs this word?

    // ...
    val management = AkkaManagement(system)
    management.setAsyncAuthenticator(myUserPassAuthenticator)
    management.start()
    ```

Java
:   ```java
    final Function<Optional<ProvidedCredentials>, CompletionStage<Optional<String>>>
      myUserPassAuthenticator = opt -> {
        if (opt.filter(c -> (c != null) && c.verify("p4ssw0rd")).isPresent()) {
          return CompletableFuture.completedFuture(Optional.of(opt.get().identifier()));
        } else {
          return CompletableFuture.completedFuture(Optional.empty());
        }
      };
    // ...
    management.setAsyncAuthenticator(myUserPassAuthenticator);
    management.start();
    ```

@@@ note
  You can combine the two security options in order to enable HTTPS as well as basic authentication.
  In order to do this, invoke both `setAsyncAuthenticator` as well as `setHttpsContext` *before* calling `start()`.
@@@

## Stopping Akka Management

In a dynamic environment you might stop instances of Akka Management, for example if you don't want to free up resources
taken by the HTTP server serving the management routes.

> You can do so by calling `stop()` on @scaladoc[AkkaManagement](akka.management.http.AkkaManagement).

Reviewer 2: No unidoc here yet?

Author: No unidoc in akka-management afaics.

    httpClusterManagement.stop();
    ```

## Developing Extensions

> This project provides a set of management extensions. If you want to write third-party extensions to Akka management here

Reviewer: ```suggestion
This project provides a set of management extensions. To write third-party extensions to Akka Management, here
```

are few pointers about how it all works together.

The `akka-management` module provides the central HTTP endpoint to which extensions can register themselves.

For example, the "Cluster HTTP Management" module exposes HTTP routes that can be used to monitor,
and even trigger joining/leaving/downing decisions via HTTP calls to these routes. The routes and
logic for these are implemented inside the `akka-management-cluster-http`.

An extension can contribute to the exposed HTTP routes by appending to the `akka.management.http.route-providers` list in
its own `reference.conf` (make sure to use `+=` instead of `=`). The core `AkkaManagement` extension

> then collects all the routes and serves them together under the management HTTP server. This is in order

Reviewer: ```management``` vs ```Management```? I don't know, just bringing it up

Author: Yeah, good point. I guess we are a bit sloppy on deciding on consistent terms and following through with them.

Author: Added https://github.com/lightbend/paradox/issues/274 to add a term/spell checker to paradox (though, that wouldn't have helped here I guess).

> then collects all the routes and serves them together under the management HTTP server. This is in order

Reviewer: ```suggestion
then collects all the routes and serves them together under the management HTTP server. This
```

> to avoid having to start an additional HTTP server for each additional extension, and also, it allows

Reviewer: ```suggestion
avoids having to start an additional HTTP server for each additional extension, and additionally allows
```

easy extension of routes served by including libraries that offer new capabilities (such as health-checks or

> cluster information etc).

Reviewer: Kind of run-on, maybe break the sentence up.

Author: Yep, reworked the whole section a bit.

> Management extensions whose nature is "active" (as in, they "do something proactively") should not be

Reviewer: ```suggestion
Management extensions whose nature is "active", e.g. they do something proactively, should not be
```

> started automatically, but should instead be started manually by the user. One example of that is the Cluster

Reviewer: ```suggestion
started automatically, but rather started manually by the user. One example of that is the Cluster
```

Bootstrap, which on one hand does contributes routes to Akka Management, however the bootstrapping process
does not start unless `ClusterBootstrap().start()` is invoked, which allows the user to decide when exactly

> it is ready and wants to start joining an existing cluster.

Reviewer: A run-on. Perhaps:
```suggestion
One example of the recommended manual operations is Cluster Bootstrap, which contributes routes to Akka Management. The bootstrapping process does not start unless `ClusterBootstrap().start()` is invoked, allowing the user to decide exactly when it is ready to start joining an existing cluster.
```
